src/sagas/reown.js

```diff
@@ -0,0 +1,861 @@
+            onComplete: (pinCode) => {
+              dispatch(hideGlobalModal());
+              pinResolve(pinCode);
+            },
+            onCancel: () => {
+              dispatch(hideGlobalModal());
+              pinReject(new Error('PIN entry cancelled'));
+            }
+          }));
+        });
+
+        pinPromise
+          .then((pinCode) => {
+            resolve({
+              type: TriggerResponseTypes.PinRequestResponse,
+              data: {
+                accepted: true,
+                pinCode,
+              }
+            });
+          })
+          .catch(() => {
+            resolve({
+              type: TriggerResponseTypes.PinRequestResponse,
+              data: {
+                accepted: false,
+                pinCode: null,
+              }
+            });
+          });
+      } break;
+
+      default:
+        reject(new Error('Invalid request'));
+    }
+  });
+
+export function* onSignMessageRequest({ payload }) {
+  const { accept: acceptCb, deny: denyCb, data, dapp } = payload;
+  const wallet = getGlobalWallet();
+
+  if (!wallet.isReady()) {
+    log.error('Got a session request but wallet is not ready.');
+    return;
+  }
+
+  yield put(showGlobalModal(MODAL_TYPES.REOWN, {
+    type: ReownModalTypes.SIGN_MESSAGE,
+    data: {
+      data,
+      dapp,
+    },
+    onAcceptAction: acceptCb,
+    onRejectAction: denyCb,
+  }));
+
+  const { deny } = yield race({
+    accept: take(types.REOWN_ACCEPT),
+    deny: take(types.REOWN_REJECT),
+  });
+
+  if (deny) {
+    denyCb();
+    return;
+  }
+
+  acceptCb();
+}
+
+export function* onSignOracleDataRequest({ payload }) {
+  const { accept, deny: denyCb, data, dapp } = payload;
+  const wallet = getGlobalWallet();
+
+  if (!wallet.isReady()) {
+    log.error('Got a session request but wallet is not ready.');
+    return;
+  }
+
+  yield put(showGlobalModal(MODAL_TYPES.REOWN, {
+    type: ReownModalTypes.SIGN_ORACLE_DATA,
+    data: {
+      data,
+      dapp,
+    },
+    onAcceptAction: accept,
+    onRejectAction: denyCb,
+  }));
+
+  const { deny } = yield race({
+    accept: take(types.REOWN_ACCEPT),
+    deny: take(types.REOWN_REJECT),
+  });
+
+  if (deny) {
+    denyCb();
+    return;
+  }
+
+  accept();
+}
+
+export function* onSendNanoContractTxRequest({ payload }) {
+  const { accept, deny: denyCb, data, dapp } = payload;
+  const wallet = getGlobalWallet();
+
+  if (!wallet.isReady()) {
+    log.error('Got a session request but wallet is not ready.');
+    return;
+  }
+
+  yield put(showGlobalModal(MODAL_TYPES.REOWN, {
+    type: ReownModalTypes.SEND_NANO_CONTRACT_TX,
+    data: {
+      data,
+      dapp,
+    },
+    onAcceptAction: accept,
+    onRejectAction: denyCb,
+  }));
+
+  const { deny } = yield race({
+    accept: take(types.REOWN_ACCEPT),
+    deny: take(types.REOWN_REJECT),
+  });
+
+  if (deny) {
+    denyCb();
+    return;
+  }
+
+  accept();
+}
+
+export function* onCreateTokenRequest({ payload }) {
```

Comment:
This and the other 3 methods above are the same, except for the `ReownModalTypes`. Is it going to be always like this? Can we reuse this code?

Reply:
Refactored, much better, thanks!